plugins/kubernetes/app/models/watchers/cluster_pod_error_watcher.rb

```diff
@@ -2,34 +2,38 @@ module Watchers
   class ClusterPodErrorWatcher < BaseClusterWatcher
     include Celluloid::Notifications
 
-    def initialize(client)
-      @client = client
-      super(@client.watch_events(field_selector: 'involvedObject.kind=Pod'))
+    attr_reader :cluster
+
+    def initialize(cluster)
+      super(cluster.client.watch_events(field_selector: 'involvedObject.kind=Pod'))
+      @cluster = cluster
     end
 
     private
 
     def handle_notice(notice)
-      if error_notice?(notice.object.reason)
-        project_name = project_name(notice.object.involvedObject)
-        publish(project_name, notice) if project_name
+      event = Events::ClusterEvent.new(notice)
+
+      if pod_error?(event)
+        pod = get_pod(event.involved_object)
+        publish("pod-events-project-#{pod.project_id}", message(event, pod: pod)) unless pod.nil?
       end
     end
 
-    def project_name(involved_object)
-      pod = @client.get_pod(involved_object.name, involved_object.namespace)
-      pod.metadata.labels ? pod.metadata.labels.project : nil
+    def get_pod(involved_object)
+      api_pod = cluster.client.get_pod(involved_object.name, involved_object.namespace)
+      Kubernetes::Api::Pod.new(api_pod)
     rescue KubeException => e
-      if e.error_code == 404
```

Comment:
The problem was that any other errors apart from 404 were causing the exception to be raised and the actor to crash. Even though it is supervised, between crashing and restarting, we were loosing events, thus, the UI was not being updated correctly. So, my thinking here is: we make an attempt to get the Pod. If it fails, we log the exception and return `nil`. Then, in the method `handle_notice`, we ignore the event if we were unable to get the Pod.

Example of an error that was causing the watcher to crash: 401 error returned by the Cluster when querying for a Pod that no longer exists, after the internal event history in Kubernetes has been wiped out (apparently, there's a treshhold in Kubernetes after which the obsolete events (events related to non-existent resources) are cleaned.

I was able to see this happening quite a lot when triggering several consecutive releases, some of them with a big number of replicas.